### PR TITLE
Refactor HTTP attributes extractors to use composition over inheritance

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
I realized that it's probably better to do that before starting the `http.route` refactorings.
HTTP client & server attributes extractors have share a common base class, so it was hard to split them into 2 PRs - that's why this one is so large. Still, ~90% of it is just renaming stuff in instrumentations.

### CODE REVIEW

File: instrumentation/ratpack/ratpack-1.7/library/src/main/java/io/opentelemetry/instrumentation/ratpack/RatpackTracingBuilder.java [hunk 1]
```diff
@@ -64,24 +67,51 @@ public RatpackTracingBuilder addClientAttributeExtractor(
    *
    * @param capturedHttpHeaders An instance of {@link CapturedHttpHeaders} containing the configured
    *     HTTP request and response names.
+   * @deprecated Use {@link #captureHttpServerHeaders(CapturedHttpHeaders)} instead.
```
Comment (trask) [inline instrumentation/ratpack/ratpack-1.7/library/src/main/java/io/opentelemetry/instrumentation/ratpack/RatpackTracingBuilder.java:70]: 👍

File: instrumentation/google-http-client-1.19/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/googlehttpclient/GoogleHttpClientSingletons.java [hunk 2]
```diff
@@ -24,23 +22,19 @@
   private static final Instrumenter<HttpRequest, HttpResponse> INSTRUMENTER;
 
   static {
-    HttpClientAttributesExtractor<HttpRequest, HttpResponse> httpAttributesExtractor =
-        new GoogleHttpClientHttpAttributesExtractor();
-    SpanNameExtractor<? super HttpRequest> spanNameExtractor =
-        HttpSpanNameExtractor.create(httpAttributesExtractor);
-    SpanStatusExtractor<? super HttpRequest, ? super HttpResponse> spanStatusExtractor =
-        HttpSpanStatusExtractor.create(httpAttributesExtractor);
+    GoogleHttpClientHttpAttributesGetter httpAttributesExtractor =
```
Comment (trask) [inline instrumentation/google-http-client-1.19/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/googlehttpclient/GoogleHttpClientSingletons.java:25]: `httpAttributesExtractor` -> `httpAttributesGetter`

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientSingletons.java [hunk 3]
```diff
@@ -23,23 +21,19 @@
...
 
   static {
-    HttpClientAttributesExtractor<RequestContext, Response> httpAttributesExtractor =
-        new AsyncHttpClientHttpAttributesExtractor();
-    SpanNameExtractor<RequestContext> spanNameExtractor =
-        HttpSpanNameExtractor.create(httpAttributesExtractor);
-    SpanStatusExtractor<RequestContext, Response> spanStatusExtractor =
-        HttpSpanStatusExtractor.create(httpAttributesExtractor);
+    AsyncHttpClientHttpAttributesGetter httpAttributesExtractor =
+        new AsyncHttpClientHttpAttributesGetter();
```
Comment (trask) [inline instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientSingletons.java:25]: `httpAttributesExtractor` -> `httpAttributesGetter`

there are several of these, worth a search for `Getter [a-zA-Z]*Extractor`

**Review by trask — APPROVED** [review]

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientSingletons.java (hunk 3 above)
Comment (breedx-splk) [inline instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientSingletons.java:25]: Probably rename instance to `httpAttributesGetter`

File: instrumentation/jaxrs-client/jaxrs-client-1.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrsclient/v1_1/JaxRsClientSingletons.java [hunk 4]
```diff
@@ -24,22 +22,17 @@
   private static final Instrumenter<ClientRequest, ClientResponse> INSTRUMENTER;
 
   static {
-    HttpClientAttributesExtractor<ClientRequest, ClientResponse> httpAttributesExtractor =
-        new JaxRsClientHttpAttributesExtractor();
-    SpanNameExtractor<? super ClientRequest> spanNameExtractor =
-        HttpSpanNameExtractor.create(httpAttributesExtractor);
-    SpanStatusExtractor<? super ClientRequest, ? super ClientResponse> spanStatusExtractor =
-        HttpSpanStatusExtractor.create(httpAttributesExtractor);
+    JaxRsClientHttpAttributesGetter httpAttributesExtractor = new JaxRsClientHttpAttributesGetter();
```
Comment (breedx-splk) [inline instrumentation/jaxrs-client/jaxrs-client-1.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrsclient/v1_1/JaxRsClientSingletons.java:25]: instance name to `httpAttributesGetter`

File: instrumentation/jaxrs-client/jaxrs-client-2.0/jaxrs-client-2.0-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrsclient/v2_0/JaxRsClientSingletons.java [hunk 5]
```diff
@@ -26,29 +24,24 @@
   private static final Instrumenter<ClientRequestContext, ClientResponseContext> INSTRUMENTER;
 
   static {
-    HttpClientAttributesExtractor<ClientRequestContext, ClientResponseContext>
-        httpAttributesExtractor = new JaxRsClientHttpAttributesExtractor();
-    SpanNameExtractor<? super ClientRequestContext> spanNameExtractor =
-        HttpSpanNameExtractor.create(httpAttributesExtractor);
-    SpanStatusExtractor<? super ClientRequestContext, ? super ClientResponseContext>
-        spanStatusExtractor = HttpSpanStatusExtractor.create(httpAttributesExtractor);
+    JaxRsClientHttpAttributesGetter httpAttributesExtractor = new JaxRsClientHttpAttributesGetter();
```
Comment (breedx-splk) [inline instrumentation/jaxrs-client/jaxrs-client-2.0/jaxrs-client-2.0-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrsclient/v2_0/JaxRsClientSingletons.java:27]: variable name getter

File: instrumentation/jaxrs-client/jaxrs-client-2.0/jaxrs-client-2.0-resteasy-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrsclient/v2_0/ResteasyClientSingletons.java [hunk 6]
```diff
@@ -27,29 +25,25 @@
   private static final Instrumenter<ClientInvocation, Response> INSTRUMENTER;
 
   static {
-    HttpClientAttributesExtractor<ClientInvocation, Response> httpAttributesExtractor =
-        new ResteasyClientHttpAttributesExtractor();
-    SpanNameExtractor<? super ClientInvocation> spanNameExtractor =
-        HttpSpanNameExtractor.create(httpAttributesExtractor);
-    SpanStatusExtractor<? super ClientInvocation, ? super Response> spanStatusExtractor =
-        HttpSpanStatusExtractor.create(httpAttributesExtractor);
+    ResteasyClientHttpAttributesGetter httpAttributesExtractor =
```
Comment (breedx-splk) [inline instrumentation/jaxrs-client/jaxrs-client-2.0/jaxrs-client-2.0-resteasy-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jaxrsclient/v2_0/ResteasyClientSingletons.java:28]: variable name getter

File: instrumentation/jetty-httpclient/jetty-httpclient-9.2/library/src/main/java/io/opentelemetry/instrumentation/jetty/httpclient/v9_2/internal/JettyClientInstrumenterBuilder.java [hunk 7]
```diff
@@ -49,22 +47,18 @@ public JettyClientInstrumenterBuilder captureHttpHeaders(
   }
 
   public Instrumenter<Request, Response> build() {
-    HttpClientAttributesExtractor<Request, Response> httpAttributesExtractor =
-        new JettyClientHttpAttributesExtractor(capturedHttpHeaders);
-    SpanNameExtractor<Request> spanNameExtractor =
-        HttpSpanNameExtractor.create(httpAttributesExtractor);
-    SpanStatusExtractor<Request, Response> spanStatusExtractor =
-        HttpSpanStatusExtractor.create(httpAttributesExtractor);
+    JettyClientHttpAttributesGetter httpAttributesExtractor = new JettyClientHttpAttributesGetter();
```
Comment (breedx-splk) [inline instrumentation/jetty-httpclient/jetty-httpclient-9.2/library/src/main/java/io/opentelemetry/instrumentation/jetty/httpclient/v9_2/internal/JettyClientInstrumenterBuilder.java:50]: variable name getter

**Review by breedx-splk — APPROVED** [review]
Thanks! This really does feel considerably cleaner to me. I called out a few consistency issues, but really looks good overall.

File: instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientSingletons.java (hunk 3 above)
Comment (breedx-splk) [reply instrumentation/async-http-client/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/v2_0/AsyncHttpClientSingletons.java:25]: doh, we probably duplicated efforts. 🤷🏻 coverage I guess. heh.

**Comment by trask** [comment]
> This really does feel considerably cleaner to me

I agree, I thought it was clearer having the multiple extractors built on top of the getter compared to built on top of another extractor

**Pull request merged**